Helper: Raw Pointer store/load

ax3l opened the pull request:
Quick & dirty draft to share raw pointers for read & write.

Related to #140

### To Do

- [x] add docstrings
- [x] add tests
- [x] loadChunk: also a `shared_ptr` data interface, separate out allocation API
- [x] move helpers into own header
- [x] ~unit tests for reference counts of `shared_ptr`s~ deferred to #211

ax3l committed:
Helper: shareRaw

Quick & dirty draft to share raw pointers for read & write.

ax3l committed:
shareRaw: tests

ax3l committed:
[draft] loadChunk

rly?

ax3l commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -29,10 +29,41 @@
+//! @{
+/** Share ownership with a raw pointer
+ *
+ * Helper function to share read/write ownership
```
write

ax3l commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -29,10 +29,41 @@
+ */
+template< typename T >
+std::shared_ptr< T >
+shareRaw( T* x )
```
possible naming: `storeRaw`

since we will need another wrapper for loads.

on the other hand, we can also use `shareRaw`/`storeRaw` in `determineDatatype` arguments.

anokfireball replied:
Adapting from future C++ lingo, we might want to call this a [```span```](http://en.cppreference.com/w/cpp/container/span) (I'd still prefer the name [```array_view```](http://open-std.org/Jtc1/sc22/wg21/docs/papers/2017/p0122r4.pdf), but well :man_shrugging:).
One could think about ```shared_span```/```unique_span``` and ```readonly_span```/```readwrite_span```

anokfireball replied:
Sidenote - ```span``` is available in [GSL](https://github.com/Microsoft/GSL/blob/master/include/gsl/span) should we want to use it.
It does pretty much the same thing as our use of ```std::unique_ptr``` and ```shared_ptr```, conveys the message that it is about a view into a range of values but does not indicate anything about ownership and lifetime.

ax3l replied:
Span name: Fits quite good as well, similar to the `ContiguousIterator` concept discussed earlier.

GSL: not sure, maybe as an overload. But could be too heavy for now (C++11), also dependency constrain wise. (uhh, [GSL-lite](https://github.com/martinmoene/gsl-lite) ;-) )

Also, I think if I want to express "safe" numpy arrays in the Python API I will need the lifetime control as well there (changing numpy.array.flags).

ax3l commented on test/CoreTest.cpp:
```diff
@@ -452,6 +452,22 @@ TEST_CASE( "structure_test", "[core]" )
+
+template< typename T, std::size_t T_size >
+std::unique_ptr< T[], nop >
+wrapRaw( std::array< T, T_size > & a )
```
`loadRaw`?

ax3l commented on test/CoreTest.cpp:
```diff
@@ -503,11 +519,13 @@ TEST_CASE( "wrapper_test", "[core]" )
     mrc2.makeConstant(value);
-    o.iterations[4].meshes["E"]["y"].loadChunk({0}, {1}, loadData);
+    std::array< double, 1 > moreData = { 112233. };
+    auto pppp = wrapRaw(moreData);
```
we need an `lvalue`

ax3l commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -29,10 +29,41 @@
+ * Helper function to share write ownership
+ * unprotected and without reference counting with a
+ * raw pointer or stdlib container.
+ */
```
add `@note` or `@warning` about user-responsibility of memory pre-allocation, buffer-consistency, etc.

ax3l commented on test/CoreTest.cpp:
```diff
@@ -503,11 +519,13 @@ TEST_CASE( "wrapper_test", "[core]" )
     mrc2.makeConstant(value);
-    o.iterations[4].meshes["E"]["y"].loadChunk({0}, {1}, loadData);
+    std::array< double, 1 > moreData = { 112233. };
+    auto pppp = loadRaw(moreData);
```
note: we need an `lvalue` since `loadChunk` needs a `ref`

we could also provide a pure `*` interface that does this wrapping.

ax3l commented on test/CoreTest.cpp:
```diff
@@ -503,11 +519,13 @@ TEST_CASE( "wrapper_test", "[core]" )
     mrc2.makeConstant(value);
-    o.iterations[4].meshes["E"]["y"].loadChunk({0}, {1}, loadData);
+    std::array< double, 1 > moreData = {{ 112233. }};
+    auto pppp = loadRaw(moreData);
```
note: we need an `lvalue` since `loadChunk` needs a `ref`

we could also provide a pure `*` interface that does this wrapping.

anokfireball replied:
As we introduce more user-friendly ways of passing raw data here, you absolutely have my blessing if you want to change the interface of ```loadChunk``` and ```storeChunk```.

ax3l replied:
I will probably just add two overloads for raw pointers then and will do the wrapping therein.
Same for `determineDatatype<T*>`.

anokfireball replied:
Just so the intention behind the ```std::shared_ptr``` does not get lost in the ether:
It is supposed to signal that the data behind it must not be changed until the reference count is back to the initial value (i.e. the backend does not hold a reference any more):

```cpp
std::shared_ptr< double > data = /* ... */
Offset offset = /* ... */
Extent extent = /* ... */
long initial = data.use_count();

auto record_component = /* ... */
record_component.storeChunk(offset, extent, global_data);
std::future< void > future = s.flush();

while( data.use_count() > initial )
{
    /* wörk that DOES NOT modify *data */
}
/* no more such wörk */
future.wait();
/* wörk that DOES modify *data */
```

ax3l replied:
Can you elaborate why it's a `shared` ptr for `stores` but a `unique` ptr for loads? (I understand the in-out meaning of `unique_ptr<T> &` and wonder why we use it only in load)

anokfireball replied:
Legacy code and laziness to refactor (or even *delete*) it, in all honesty:
https://github.com/openPMD/openPMD-api/blob/62b854a0714e4f6516bc5515f0a4382e673a6b56/include/openPMD/RecordComponent.hpp#L162-L164
The previous interface induced a ```flush()``` and *returned* the pointer, eliminating the need for shared ownership. As the ```unique_ptr``` is more lightweight, it was chosen for that purpose. *This might not have been premature optimization, but now is overly specific.*

ax3l replied:
So does it make sense to put both `unique_ptr<T>&` or both `shared_ptr<T>`?

anokfireball replied:
[Semantic obviousness](https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/) for the user would dictate a ```shared_ptr<T>```. Note that this is also used internally when writing datasets (to take advantage of reference-counting):
https://github.com/openPMD/openPMD-api/blob/62b854a0714e4f6516bc5515f0a4382e673a6b56/include/openPMD/IO/IOTask.hpp#L162-L169
The downside with ```unique_ptr<T>&``` is that the original ```unique_ptr<T>``` might go out of scope for the user, destroy and deallocate what it points to. The reference would be dangling, and we cannot guarantee the IOTask has been completed by that point:
```cpp
auto record_component = /* ... */
{
    std::unique_ptr< double > data;
    Offset offset = /* ... */
    Extent extent = /* ... */
    record_component.loadChunk(offset, extent, data);
}
s.flush();
```
With a ```shared_ptr<T>```, the operation would be wasteful as the results are discarded. But at least the program will not explode in your face, thanks to reference-counting the runtime won't deallocate the pointed-to-data.

**NOTE - This is what the behaviour should be, but currently is not:**
For some reason (maybe the above?) I implemented ```loadChunk``` to always [```.flush()```](https://github.com/openPMD/openPMD-api/blob/069a9e2280c3e501007b5b164c27f5dcc375f29f/include/openPMD/RecordComponent.hpp#L195), but ```storeChunk``` to defer flushes with the help of a [queue](https://github.com/openPMD/openPMD-api/blob/069a9e2280c3e501007b5b164c27f5dcc375f29f/include/openPMD/RecordComponent.hpp#L277). Ideally, we would put both into the queue.

**MINOR NOTE** With similar logic, there should not be any need to [```.flush()```](https://github.com/openPMD/openPMD-api/blob/d16f1762fcb1be7e7441755a3209abdba7994173/include/openPMD/RecordComponent.hpp#L158) for every IOTask generated by ```storeChunk```.

ax3l replied:
Nice, I agree with your points and the ownership arguments in the reference.
I think all in all it makes sense then to make `loadChunk` use a `shared_ptr` too.

ax3l replied:
Should we then overload both methods with a "lazy"/"unaware" raw-pointer overload or just expose the `shareRaw` wrapper to the user?

anokfireball replied:
What are use cases you're thinking of for raw pointers?
Even with such applications, I would rather keep the contract (i.e. shared ownership) in the interface (i.e. offer ```loadChunk```/```storeChunk``` with ```shared_ptr``` only) and offer a helper for raw pointers:
```cpp
template< typename T >
std::shared_ptr< T >
storeRaw( T* ptr )
{
    return std::shared_ptr< T >( ptr, [](T*){} );
}
```

anokfireball replied:
This helper mechanism does not offer the safety-guarantees I referred to in my [comment](https://github.com/openPMD/openPMD-api/pull/204#discussion_r190864021), btw:
```cpp
auto record_component = /* ... */
{
    std::vector< double > data = /* ... */
    Offset offset = /* ... */
    Extent extent{data.size()};
    record_component.storeChunk(offset, extent, storeRaw(data));
}
//pointer behind `data` does not exist any more,
//but flush might not have happened
s.flush();
```

ax3l replied:
> What are use cases you're thinking of for raw pointers?

Mainly contiguous containers (STL and third-party libs), that don't offer a shared view or unique-ptr into their data.

> I would rather keep the contract (i.e. shared ownership) in the interface

Agreed, so we will expose the helpers/wrappers and do not overload.

ax3l replied:
While refactoring the `loadChunk` interface to `shared_ptr` I realized the following: For proper support of the "we allocate for you" API I would need to pass as `shared_ptr< T[] >&` which I find a little odd. Also, the implementation allocated immediately which does not offer too much convenience.

What do you think about if I:
- remove the allocation policy argument
- use a signature (1) that returns a `shared_ptr<T>` (we allocate)
- use a signature (2) that accepts a pre-allocated `shared_ptr<T>` (user allocates)

I think that makes things clearer.

anokfireball replied:
Overall I'll have to agree. Didn't think far enough to realize that you can't allocate the ```shared_ptr<T>``` and have it visible to the caller. The proposal with the two signatures also makes sense.
Do you have any catchy names for semantic differentiation (i.e. ```loadChunkAllocated```)? 

Quick idea that does not entail two functions:
```cpp
template< typename T, typename D >
std::shared_ptr< T, typename D >
loadChunk(Offset const&,
          Extent const&,
          std::shared_ptr< T, D > = std::shared_ptr< T, D >()); // assume allocated if not default
          //double targetUnitSI = std::numeric_limits< double >::quiet_NaN() );
```

> use a signature (1) that returns a shared_ptr<T> (we allocate)

Good thing I was lazy enough to leave that code in there.

ax3l replied:
I don't think we will need different names, the calling parameters are different :-)

> Quick idea that does not entail two functions.

Tried that, does not really work great also because `shared_ptr< T[] >` machinery was kind of forgotten in C++11... I can push you the draft in 6dc5ab956ba3fcec7016a4e17ac1b45f5e0800fe that RT crashes. But maybe I screwed up the deleter.

Anyway, I have the feeling that member function has too many parameters and does too little (just allocate immediately) that can not be separated otherwise cleanly, which makes it easier to grasp by being less complex.

We can implement this by the existing signature and the second just allocates, calls it and returns the new memory.

ax3l replied:
Btw, I think [this example](https://github.com/openPMD/openPMD-api/pull/204#discussion_r190829625) you posted would make a great page in the manual! :)

anokfireball replied:
> ```shared_ptr< T[] >``` machinery was kind of forgotten in C++11

How this is only coming in C++17 is beyond me...

ax3l committed:
Refactor load: shared_ptr

Refactor the `loadChunk` method to also exchange data
as shared_ptr.

Move `shareRaw()` wrapper helper to own header.

ax3l committed:
Refactor load: shared_ptr two sig

Add two separate signatures for `loadChunk`: with and without
allocation for the user.

anokfireball commented on include/openPMD/auxiliary/ShareRaw.hpp:
```diff
@@ -0,0 +1,65 @@
+    std::shared_ptr< T >
+    shareRaw( std::vector< T > & v )
+    {
+        return std::shared_ptr< T >( v.data(), [](T*){} );
```
Maybe we should SFINAE and do this for anything that offers ```T* TYPE::data()```?

ax3l replied:
That's a good idea, but probably offering a trait is better.

`*mumble*` Although the concept of `SFINAE`s and `enable_if` are slick, they are quite compiler-dependent to work well in the corner-cases (msvc #182, nvcc) and are [super-slow in terms of compile time](https://www.reddit.com/r/cpp/comments/6gur2x/the_rule_of_chiel_aka_compiletime_cost_of/).

Well, actually: just because something has a `.data()` member does not mean it is contiguous and save to use. let us not do that by default as long as we can't check it via traits.

anokfireball replied:
> just because something has a .data() member does not mean it is contiguous and save to use

My interpretation of [```std::data()```](http://en.cppreference.com/w/cpp/iterator/data) is just that:
> ... a pointer to the block of memory containing the elements ...

[xtensor](http://xtensor.readthedocs.io/en/latest/api/xcontainer.html#_CPPv3N2xt10xcontainer4dataEv) does it as well, btw.
But that is C++17, so well have to deal with it manually, anyway.

ax3l replied:
Interesting, didn't know `std::data()` yet! I stand corrected.

anokfireball commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -170,7 +134,10 @@ RecordComponent::loadChunk(Offset const& o, Extent const& e, std::unique_ptr< T[
-        data = std::unique_ptr< T[], D >(new T[numPoints]);
+    {
+        auto newData = std::shared_ptr<T>( new T[numPoints], []( T *p ){ delete [] p; } );
+        data.swap(newData);
```
Are you doing this for safety guarantee when the allocation would fail?

ax3l replied:
I was trying to exchange the incoming shared_ptr but things did not work as expected and I started to go for the two-signature option for allocations...

ax3l commented on examples/2_read_serial.cpp:
```diff
@@ -60,8 +60,7 @@ int main()
     Extent chunk_extent = {2, 2, 1};
-    std::unique_ptr< double[] > chunk_data;
-    E_x.loadChunk(chunk_offset, chunk_extent, chunk_data);
+    auto chunk_data = E_x.loadChunk<double>(chunk_offset, chunk_extent);
```
I really wish the syntax `shared_ptr< auto > = someFunc< double >();` would exist...

ax3l replied:
should I write `shared_ptr<double>` here so users know what is going on?

anokfireball replied:
As this is only the second piece of code a user would see, this is a good point in time to show them the way we handle raw data.

ax3l commented:
@C0nsultant Besides possible overloads for additional `.data()` providing containers, what do you think of the pushed changes?

anokfireball commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -123,17 +141,13 @@ RecordComponent::loadChunk(Offset const& o, Extent const& e, std::unique_ptr< T[
+    if( !data )
+        throw std::runtime_error("Unallocated pointer passed during chunk loading.");
 
     uint64_t numPoints = 1u;
```
This could now be pulled into a lower scope inside the ```if( *m_isConstant )``` branch.

C0nsultant replied:
Lines https://github.com/ax3l/openPMD-api/blob/fd7b2b1b14e8b574c3b35d5ef00bb50f59441926/include/openPMD/RecordComponent.hpp#L176-L226 can now safely be removed as the functionality is provided now.

anokfireball reviewed (approved):
Love the new syntax!
```cpp
auto chunk_data = E_x.loadChunk<double>(chunk_offset, chunk_extent);
```
is what the initial draft was and *now it's back*!

anokfireball commented:
Lines https://github.com/ax3l/openPMD-api/blob/fd7b2b1b14e8b574c3b35d5ef00bb50f59441926/include/openPMD/RecordComponent.hpp#L176-L226 can now safely be removed as the functionality is provided now.

ax3l commented:
found a little design flaw. `writeChunk` should only take `shared_ptr<T const>` args. will fix on the way :)
(discovered while trying to write a `double const`)

update: I was overly optimistic, can only fix the internal `Parameter` type.

ax3l commented:
@C0nsultant this should now contain all changes :)

anokfireball commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -124,24 +140,16 @@ RecordComponent::loadChunk(Offset const& o, Extent const& e, std::shared_ptr< T
+        uint64_t numPoints = 1u;
+        for( auto const& dimensionSize : e )
+            numPoints *= dimensionSize;
         /*
```
This block of code can also be removed. The constant value is cached within the ```RecordComponent``` itself now.

anokfireball commented:
By making ```loadChunk``` defer actual loading, we also need to introduce proper reference counting for the ```shared_ptr```, just like with ```storeChunk```.

This *probably* (haven't thought about it for long enough) means we have to make [this](
https://github.com/ax3l/openPMD-api/blob/23e95403e6bb212e83e2929b969eabf10dd7cef8/include/openPMD/IO/IOTask.hpp#L177) raw pointer into a ```shared_ptr< void >``` as well. Otherwise the same out-of-scope/deallocation problem could hit us during loading.

**EDIT**
Yes, I am fairly certain this needs to happen:
```cpp
auto record_component = /* ... */
{
    std::shared_ptr< double > data = record_component.loadChunk(offset, extent);
    //data.ref_count() == 1
}
//the IOTask in the backend still holds the raw pointer to the now deallocated memory behind data
s.flush();
```

Do you want me to do that, @ax3l? I'll push it to this PR. Just making sure we're not messing with the same branch simulateneously.

anokfireball committed:
Modify IOTask parameter accoringly

Have Parameter< Operation::READ_DATASET > now also carry a shared_ptr
for correct reference counting.
Refactor PatchRecordComponent to reflect the same changes as
RecordComponent.

ax3l commented:
good point, feel free to go ahead :)
(maybe also add this ref count as an `REQUIRE` in a test? I missed that we should check that indeed)

ax3l commented:
@C0nsultant what I am quite surprised is that we are just doing

```C++
auto data = record_component.loadChunk(offset, extent);
// use data
```

without the need to call `flush()` first. Is this intended? Shouldn't the data first be available *after* `flush()`?

Side note: what happened to the CI in this PR? Somehow it's doubled since yesterday evening.

anokfireball commented:
> Shouldn't the data first be available after ```flush()```?

You are entirely correct. *In theory...* This is not the current behaviour, but will be introduced with #205.

> what happened to the CI in this PR?

Not sure, this happened after I pushed to your branch (following the steps GitHub's documentation lists).

ax3l commented:
All right then, let's wrap it :-)
Thanks for the push!
